src/walberla_bridge/src/lattice_boltzmann/LBWalberlaImpl.hpp

```diff
@@ -633,45 +623,37 @@ class LBWalberlaImpl : public LBWalberlaBase {
       (*m_boundary)(&block);
   }
 
-  void integrate_push_scheme() {
+  void integrate_update_velocities_from_pdf(
+      std::shared_ptr<BlockStorage> const &blocks) {
+    for (auto b = blocks->begin(); b != blocks->end(); ++b)
+      (*m_update_velocities_from_pdf)(&*b);
+  }
+
+  void integrate_pull_scheme() {
     auto const &blocks = get_lattice().get_blocks();
     // Reset force fields
     integrate_reset_force(blocks);
-    // LB collide
-    integrate_collide(blocks);
-    m_pdf_streaming_communicator->communicate();
-    // Handle boundaries
-    if (m_has_boundaries) {
-      integrate_boundaries(blocks);
-    }
-    // LB stream
-    integrate_stream(blocks);
+    // LB stream collide
+    integrate_stream_collide(blocks);
     // Mark pending ghost layer updates
-    m_pending_ghost_comm.set(GhostComm::PDF);
+    // As pdf and laf are communicated direcly afterwards they are not set.
     m_pending_ghost_comm.set(GhostComm::VEL);
     m_pending_ghost_comm.set(GhostComm::LAF);
-    // Refresh ghost layers
```

Comment:
As I understand it, the last applied force is communicated by the pdf_streaming_communicator. If this is the case, LAF should probably not be marked as pending, here.

Reply:
Yes, however, there are some unit tests (LBWalberlaImpl_unit_tests, LBWalberlaImpl_ bspline_tests) which need the flag to be set to true after the integration step, so that the forces can be communicated. it would probably be better to have function, which can communicate regardless of the flag.